### PR TITLE
Added Salesforce REST api Connector

### PR DESCRIPTION
**Pull request by hariss63**

### CODE REVIEW

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        <connector.version>1.0.0</connector.version>
+    </properties>
+    <dependencies>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.mediation.initializer</artifactId>
+            <version>4.2.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.synapse</groupId>
```
Comment (Shakila) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml]: This dependency is already added in parent pom. Can we remove from here?

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+                                </resource>
+                            </resources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+    <repositories>
+        <repository>
```
Comment (Shakila) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml]: This repository also added in parent pom.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+               description="Value of your secret key given when you registered your application with salesforce."/>
+    <parameter name="clientId"
+               description="Value of your API Key given when you registered your application with salesforce."/>
+    <sequence>
+        <property name="uri.var.apiVersion" expression="$func:apiVersion"/>
+        <property name="uri.var.hostName" expression="$func:hostName"/>
+        <property name="uri.var.refreshToken" expression="$func:refreshToken"/>
+        <property name="uri.var.clientSecret" expression="$func:clientSecret"/>
+        <property name="uri.var.clientId" expression="$func:clientId"/>
+        <property name="uri.var.accessToken"
```
Comment (malakasilva) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml]: This is wrong user needs to pass the access token for the first request. Also if he does not want the access token to be refreshed we should handle it. According to this access token needs to be there in registry first.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                  expression="get-property('registry', 'connectors/salesforceRest/accessToken')"/>
+        <property name="systemTime" expression="get-property('SYSTEM_TIME')"
+                  scope="default"/>
+        <property name="registryTime"
+                  expression="get-property('registry', 'connectors/salesforceRest/time')"/>
```
Comment (malakasilva) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml]: Registry path should be configurable by user.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        <property name="uri.var.accessTokenRegistryPath" expression="$func:accessTokenRegistryPath"/>
+        <property name="uri.var.timeRegistryPath" expression="$func:timeRegistryPath"/>
+        <property name="uri.var.apiUrlRegistryPath" expression="$func:apiUrlRegistryPath"/>
+
+        <filter source="boolean($ctx:uri.var.accessToken)" regex="false">
+            <property name="uri.var.accessToken"
+                      expression="get-property('registry', $ctx:uri.var.accessTokenRegistryPath)"/>
+        </filter>
+
+        <filter source="boolean($ctx:uri.var.accessTokenRegistryPath)" regex="false">
```
Comment (malakasilva) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml]: Please move this as the first step.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            <then>
+                <property name="uri.var.timeRegistryPath" value="connectors/salesforceREST/time"/>
+
+            </then>
+            <else>
+                <property name="uri.var.timeRegistryPath"
+                          expression="$ctx:uri.var.timeRegistryPath"/>
+            </else>
+        </filter>
+        <filter source="boolean($ctx:uri.var.apiUrlRegistryPath)" regex="false">
```
Comment (malakasilva) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml]: No need to do this user always has to send it

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                          expression="get-property('registry', $ctx:uri.var.timeRegistryPath)"/>
+                <script language="js"><![CDATA[
+                    var time1 = mc.getProperty("systemTime");
+                    var time2 = mc.getProperty("registryTime");
+                    var timeTaken = time1 - time2;
+                    mc.setProperty("INTERVAL_TIME", timeTaken);
+                   ]]>
+                </script>
+                <!--If the access token is not in the registry(first call) or time interval less than 50 minutes , then get a new access token by calling getAccessTokenByRefreshToken method and store the time into registry-->
+                <filter source="$ctx:INTERVAL_TIME &gt;3000000 or $ctx:uri.var.accessToken=''"
```
Comment (malakasilva) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/init.xml]: Interval time needs to come from user as an optional param.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/getAccessTokenFromRefreshToken.xml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                  mc.getConfiguration().getRegistry().updateResource(newTimeRegistryPath, newSystemTime);
+                }
+             }
+         ]]>
+        </script>
+        <!-- Remove custom Headers from the API Response -->
+        <header name="Cache-Control" action="remove" scope="transport"/>
+        <header name="Pragma" action="remove" scope="transport"/>
+    </sequence>
+</template>
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/getAccessTokenFromRefreshToken.xml]: Remove empty lines

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-search/component.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        </component>
+        <component name="searchSuggestedQueries">
+            <file>searchSuggestedQueries.xml</file>
+            <description>Returns a list of suggested searches based on the user’s query string text
+                matching searches that other users have performed in Salesforce Knowledge
+            </description>
+        </component>
+    </subComponents>
+</component>
+
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-search/component.xml]: Remove empty lines

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-users/component.xml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+            <file>getUserInformation.xml</file>
+            <description>Get User Information From Salesforce.</description>
+        </component>
+        <component name="resetPassword">
+            <file>resetPassword.xml</file>
+            <description>Reset Password for Salesforce account for a specific User.</description>
+        </component>
+    </subComponents>
+</component>
+
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-users/component.xml]: Remove empty lines

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,830 @@
+package org.wso2.carbon.connector.integration.test.salesforcerest;
+/*
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java]: Shall we move licence header to top?

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/AXIS2/config/test_axis2_server_9000.xml [hunk 12]
```diff
@@ -0,0 +1,378 @@
+<!--
+  ~  Licensed to the Apache Software Foundation (ASF) under one
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/AXIS2/config/test_axis2_server_9000.xml]: Wrong license header/template. Please use WSO2 template.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/ESB/samples/synapse_sample_0.xml [hunk 13]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/ESB/samples/synapse_sample_0.xml]: Why we need synapse samples here?

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/getAccessTokenFromRefreshToken.xml (hunk 8 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-config/getAccessTokenFromRefreshToken.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-search/component.xml (hunk 9 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-search/component.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-users/component.xml (hunk 10 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-users/component.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java (hunk 11 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/AXIS2/config/test_axis2_server_9000.xml (hunk 12 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/AXIS2/config/test_axis2_server_9000.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/ESB/samples/synapse_sample_0.xml (hunk 13 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/resources/artifacts/ESB/samples/synapse_sample_0.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.mediation.initializer</artifactId>
+            <version>4.2.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon</groupId>
+            <artifactId>org.wso2.carbon.connector.core</artifactId>
+            <version>4.2.0</version>
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml]: Use properties to get the values to versions.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml [hunk 15]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml]: Add licence header to all files.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+<assembly>
+    <formats>
+        <format>zip</format>
+    </formats>
+
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml]: Remove empty lines.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-quickaction/component.xml [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            <description>Return a specific action’s descriptive detail</description>
+        </component>
+        <component name="getDefaultValueOfAction">
+            <file>getDefaultValueOfAction.xml</file>
+            <description>Return a specific action’s default values, including default field values
+            </description>
+        </component>
+    </subComponents>
+</component>
+
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-quickaction/component.xml]: Remove empty lines.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-records/component.xml [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        </component>
+        <component name="retrieveStandardFieldValuesFromExternalObjectWithExternalId">
+            <file>retrieveStandardFieldValuesFromExternalObjectWithExternalId.xml</file>
+            <description>Retrieve specific Standard field values from an External Object using
+                External Id
+            </description>
+        </component>
+    </subComponents>
+</component>
+
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-records/component.xml]: Remove empty lines.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,830 @@
...
+import java.io.IOException;
+import java.lang.String;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SalesforceRestConnectorIntegrationTest extends ConnectorIntegrationTestBase {
+    private Map<String, String> esbRequestHeadersMap = new HashMap<String, String>();
+
+    private Map<String, String> apiRequestHeadersMap = new HashMap<String, String>();
+
```
Comment (hmrajas) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java]: Remove empty lines.

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml (hunk 14 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/pom.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559/commits/f11f75546b14c6830220ea4ae9684da0dab1ab3f

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml (hunk 16 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559/commits/f11f75546b14c6830220ea4ae9684da0dab1ab3f

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java (hunk 19 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/test/java/org/wso2/carbon/connector/integration/test/salesforcerest/SalesforceRestConnectorIntegrationTest.java]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559/commits/f11f75546b14c6830220ea4ae9684da0dab1ab3f

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml (hunk 15 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/assembly/assemble-connector.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-quickaction/component.xml (hunk 17 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-quickaction/component.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

File: salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-records/component.xml (hunk 18 above)
Comment (hariss63) [inline salesforcerest/salesforcerest-connector/salesforcerest-connector-1.0.0/org.wso2.carbon.connector/src/main/resources/salesforcerest-records/component.xml]: Fixed in https://github.com/wso2-attic/esb-connectors/pull/559

**Pull request merged**